src/Type/Doctrine/Query/QueryResultTypeBuilder.php

```diff
@@ -0,0 +1,111 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Doctrine\Query;
+
+use PHPStan\Type\Constant\ConstantArrayTypeBuilder;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\VoidType;
+
+/**
+ * QueryResultTypeBuilder helps building the result type of a query
+ *
+ * Like Doctrine\ORM\Query\ResultSetMapping, but for static typing concerns
+ */
+final class QueryResultTypeBuilder
+{
+
+	/** @var bool */
+	private $selectQuery = false;
+
+	/** @var array<array-key,Type> */
+	private $entities = [];
+
+	/** @var array<array-key,Type> */
+	private $scalars = [];
+
+	public function setSelectQuery(): void
+	{
+		$this->selectQuery = true;
+	}
+
+	public function isSelectQuery(): bool
+	{
+		return $this->selectQuery;
+	}
+
+	/**
+	 * @param array-key $alias
+	 */
+	public function addEntity($alias, Type $type): void
+	{
+		$this->entities[$alias] = $type;
+	}
+
+	/**
+	 * @return array<array-key,Type>
+	 */
+	public function getEntities(): array
+	{
+		return $this->entities;
+	}
+
+	/**
+	 * @param array-key $alias
+	 */
+	public function addScalar($alias, Type $type): void
+	{
+		$this->scalars[$alias] = $type;
+	}
+
+	/**
+	 * @return array<int,Type>
+	 */
+	public function getScalars(): array
+	{
+		return $this->scalars;
+	}
+
+	public function getResultType(): Type
+	{
+		// It makes no sense to speak about a result type for UPDATE or DELETE
+		// queries, so we return VoidType here.
+		if (!$this->selectQuery) {
+			return new VoidType();
+		}
```

Comment:
The thing is that the return type of getResult() or getSingleResult() is inconsistent between SELECT and UPDATE/DELETE queries.

For SELECT queries, the return types are:

```
getResult(): array<TResult>
getSingleResult(): TResult
```

For UPDATE/DELETE queries, the return types are:

```
getResult(): int
getSingleResult(): int
```

Because of this we can not use `TResult` directly in `@return`.

Instead, we will need a return type extension. However, the return type extension needs a way to find whether the `Query` instance is a SELECT or UPDATE/DELETE query. Setting TResult to void is one way to do that.

Reply:
I see. 👍

Reply:
One alternative would have been to have multiple type params on Query, so that we could do this:

```
/** 
 * @template TResult
 * @template TSingleResult
 */
class Query {
    getResult(): TResult
    getSingleResult(): TSingleResult
}
```

However I think that it would be tedious to declare the two types in annotations, like `@return Query<array<array{foo: int, bar: string}>, array{foo: int, bar: string}>`, especially because of the redundancy.

Also, we still need a return type extension so that we revert to `mixed` if the hydration mode may not be HYDRATE_OBJECT

Reply:
Yeah, I don't like this dual template type approach either.